app/services/notion.py

```diff
@@ -442,6 +460,82 @@ def build_page_url(page: Dict[str, Any]) -> Optional[str]:
     return f"{base_url}{without_hyphen}"
 
 
+async def _build_breadcrumbs(
+    entity: Dict[str, Any],
+    entity_type: str,
+    headers: Dict[str, str],
+) -> List[Dict[str, Any]]:
+    """Build breadcrumb trail from root -> current entity.
+
+    Each breadcrumb item contains: {id, type, title, url}.
+    Stops when reaching a workspace parent.
+    """
+
+    def to_node(data: Dict[str, Any], type_hint: str) -> Dict[str, Any]:
+        node_id = data.get("id")
+        title = extract_title(data, "page" if type_hint == "page" else "database")
+        url = data.get("url")
+        if type_hint == "page" and not url:
+            url = build_page_url(data)
+        return {
+            "id": node_id,
+            "type": type_hint,
+            "title": title,
+            "url": url,
+        }
+
+    chain: List[Dict[str, Any]] = []
+
+    # Walk upwards from current entity to the root
+    current = entity
+    current_type = entity_type
+    seen: Set[str] = set()
+    depth_limit = 25
+
+    for _ in range(depth_limit):
+        node_id = current.get("id")
+        if isinstance(node_id, str):
+            if node_id in seen:
+                break
+            seen.add(node_id)
+
+        chain.append(to_node(current, current_type))
+
+        parent = current.get("parent")
+        if not isinstance(parent, dict):
+            break
+        if parent.get("workspace") is True:
+            break
+
+        next_type: Optional[str] = None
+        next_id: Optional[str] = None
+        if "page_id" in parent:
+            next_type = "page"
+            next_id = parent.get("page_id")
+        elif "database_id" in parent:
+            next_type = "database"
+            next_id = parent.get("database_id")
+
+        if not next_type or not next_id:
+            break
+
+        try:
+            response = await _request_with_retry(
+                "GET",
+                f"{NOTION_API_URL}/{next_type}s/{next_id}",
+                headers=headers,
+                params=None,
+            )
+            current = response.json()
+            current_type = next_type
+        except Exception:
+            break
+
+    # Reverse to make it root -> ... -> current
+    chain.reverse()
```

Comment:
Maybe we could use `collections.deque` here instead. With `appendleft` method, we wouldn’t need to reverse the chain anymore